sdk/src/trace/batch_span_processor.cc

```diff
@@ -218,48 +228,58 @@ void BatchSpanProcessor::Export()
     }
     else
     {
-      std::weak_ptr<SynchronizationData> synchronization_data_watcher = synchronization_data_;
-      exporter_->Export(
-          nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()),
-          [notify_force_flush, synchronization_data_watcher](sdk::common::ExportResult result) {
-            // TODO: Print result
-            if (synchronization_data_watcher.expired())
-            {
+      size_t id = kInvalidExportId;
+      {
+        std::unique_lock<std::mutex> lock(synchronization_data_->async_export_data_m);
+        synchronization_data_->async_export_waker.wait_for(lock, schedule_delay_millis_, [this] {
+          return export_data_storage_->export_ids.size() > 0;
+        });
+        if (export_data_storage_->export_ids.size() > 0)
+        {
+          id = export_data_storage_->export_ids.front();
+          export_data_storage_->export_ids.pop();
+          export_data_storage_->export_ids_flag[id - 1] = false;
+        }
+      }
+      if (id != kInvalidExportId)
+      {
+        std::weak_ptr<ExportDataStorage> export_data_watcher = export_data_storage_;
+
+        std::weak_ptr<SynchronizationData> synchronization_data_watcher = synchronization_data_;
+        exporter_->Export(
+            nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()),
+            [notify_force_flush, synchronization_data_watcher, export_data_watcher,
+             id](sdk::common::ExportResult result) {
+              // TODO: Print result
+              if (synchronization_data_watcher.expired())
```

Comment:
Seems I am missing something. When will the weak_ptr to synchronization_data_ and export_data_storage_ be expired?

Reply:
These callback may be moved to another thread by another exporter, an in this situation, processor may be destroyed before invoking callback.

Reply:
ok, thanks for the explanation. Makes sense.